[WebGPU] Shader validation routines potentially rely on UB in C++

mwyrzykowski opened the pull request:
#### 655ee4d2796c44db41552773349f7cd950fcca56
<pre>
[WebGPU] Shader validation routines potentially rely on UB in C++
<a href="https://bugs.webkit.org/show_bug.cgi?id=288726">https://bugs.webkit.org/show_bug.cgi?id=288726</a>
<a href="https://rdar.apple.com/145760120">rdar://145760120</a>

Reviewed by Cameron McCormack.

The shader was using a buffer of sizeof(S) + sizeof(uint)
and writing to &amp;S::lastUintMember + 1 assuming it would
write to the uint.

This is flagged by sanitizers, so be explict and define
the member we are writing to.

* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::Buffer):
(WebGPU::Buffer::makeIndexIndirectBuffer): Deleted.
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::checkForIndirectDrawDeviceLost):

Canonical link: <a href="https://commits.webkit.org/291576@main">https://commits.webkit.org/291576@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e37141be56d89514dda28b1789b32368717f9886

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93247 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12803 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98246 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43773 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13087 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21256 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71260 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28660 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96249 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9818 "Found 1 new test failure: fast/forms/ios/focus-input-via-button-no-scaling.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84335 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51594 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9511 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43087 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79800 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/1971 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100275 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20299 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14857 "Found 1 new test failure: ipc/large-vector-allocate-failure-crash.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80280 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20551 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79596 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19790 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24143 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1462 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13414 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20283 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25460 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19970 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23430 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21711 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -166,11 +166,6 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, WGPUBufferUsageFl
 
 WTF_MAKE_TZONE_ALLOCATED_IMPL(Buffer);
 
-id<MTLBuffer> Buffer::makeIndexIndirectBuffer()
```
this function wasn't used so I removed it since it had the ` + sizeof(uint32_t)`

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -81,6 +81,17 @@ using GPUShaderValidation = MTLShaderValidation;
 using GPUShaderValidation = uint32_t;
 #endif
 
+#define WEBKIT_DRAW_INDIRECT_STRUCT_TYPES \
```
the define in a header file seems preferable to potentially conflicting definitions in C++ and MSL

mwyrzykowski commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -81,6 +81,17 @@ using GPUShaderValidation = MTLShaderValidation;
+    MTLDrawIndexedPrimitivesIndirectArguments args;         \
+    uint lostOrOOBRead;                                     \
+}; \
+struct WebKitMTLDrawPrimitivesIndirectArguments { \
```
the style-checker didn't like more than one space between the `{` and the `\`

heycam commented on Source/WebGPU/WebGPU/Device.h:
```diff
@@ -81,6 +81,17 @@ using GPUShaderValidation = MTLShaderValidation;
+#define WEBKIT_DRAW_INDIRECT_STRUCT_TYPES \
+struct WebKitMTLDrawIndexedPrimitivesIndirectArguments { \
+    MTLDrawIndexedPrimitivesIndirectArguments args;         \
+    uint lostOrOOBRead;                                     \
```
Nit: `uint` seems like a rarely used typedef, and isn't common in WebKit code. `unsigned` is more common. (And: there's not any risk of this type being a different size in the Metal program than it is on the CPU, is there? As long as we're not enabled on watchOS it's probably fine? Still, Metal framework defined structs that are shared tend to use explicitly sized integral types.)

heycam replied:
I guess `uint` is the shortest way to type `unsigned int` that works in both Metal and C++. Regardless, `uint32_t` might be better.

mwyrzykowski replied:
We can also use `unsigned` in MSL, I will use that instead

mwyrzykowski replied:
`uint` is typedef in `metal_types` as:
```
typedef unsigned int uint;
```
but the WGSL parser generates `unsigned` in places, these are all guaranteed to be the same sizes in both C++ and MSL, for consistency I think `unsigned` or `uint32_t` makes more sense as suggested

heycam commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -745,24 +748,27 @@ static void setOwnerWithIdentity(id<MTLResourceSPI> resource, auto webProcessID)
 #define primitiveRestart 1
     using namespace metal;
-    [[vertex]] void vsUshortIndexClamp(device const ushort* indexBuffer [[buffer(0)]], device MTLDrawIndexedPrimitivesIndirectArguments& indexedOutput [[buffer(1)]], const constant uint* data [[buffer(2)]], uint indexId [[vertex_id]])
+    )"  OBJC_STRINGIFY(WEBKIT_DRAW_INDIRECT_STRUCT_TYPES)   @R"(
```
Should we split the macro into two? Looks like we only need one of the structs in each of the shader programs.

heycam commented on Source/WebGPU/WebGPU/Buffer.mm:
```diff
@@ -183,9 +178,9 @@ static MTLStorageMode storageMode(bool deviceHasUnifiedMemory, WGPUBufferUsageFl
 #endif
 {
     if (m_usage & WGPUBufferUsage_Indirect)
-        m_indirectBuffer = device.safeCreateBuffer(sizeof(MTLDrawPrimitivesIndirectArguments) + sizeof(uint32_t), MTLStorageModeShared);
```
I'm not sure if we have many of these types of MTLBuffers where it just holds a single instance of a C++ struct, like these two. But it makes me wonder whether we should have a more strongly typed wrapper around `MTLBuffer`, which would be responsible for the `sizeof` calls and `static_cast`ing. Where here you'd do something like:

```
m_indirectBuffer = device.createTypedBuffer<WebKitMTLDrawPrimitivesIndirectArguments>();
```

and the wrapper class has an `operator*` etc.

Actually you could have something similar for the variable length buffers that we create too.

The more that we can encapsulate buffer size calculation, casting, pointer arithmetic, etc. into helpers, the better, and would probably reduce the amount of boilerplate when dealing with buffers that we share with the GPU.

mwyrzykowski replied:
This is a good idea. At the moment the only other known sized MTLBuffer created by WebGPU.framework is the size of a uint32_t in RenderBundleEncoder.mm - `id<MTLBuffer> outOfBoundsRead = device.safeCreateBuffer(sizeof(uint32_t));`

mwyrzykowski replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=288956 to address this

mwyrzykowski committed:
[WebGPU] Shader validation routines potentially rely on UB in C++
https://bugs.webkit.org/show_bug.cgi?id=288726
rdar://145760120

Reviewed by Cameron McCormack.

The shader was using a buffer of sizeof(S) + sizeof(uint)
and writing to &S::lastUintMember + 1 assuming it would
write to the uint.

This is flagged by sanitizers, so be explict and define
the member we are writing to.

* Source/WebGPU/WebGPU/Buffer.h:
* Source/WebGPU/WebGPU/Buffer.mm:
(WebGPU::Buffer::Buffer):
(WebGPU::Buffer::makeIndexIndirectBuffer): Deleted.
* Source/WebGPU/WebGPU/Device.h:
* Source/WebGPU/WebGPU/Device.mm:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::clampIndexBufferToValidValues):
(WebGPU::checkForIndirectDrawDeviceLost):

Canonical link: https://commits.webkit.org/291576@main
